src/function.cpp

```diff
@@ -59,32 +55,82 @@ DeclaredFunction DeclaredFunction::declare(std::vector<mlir::Type> &&domain,
   transform(domain.cbegin(), domain.cend(), back_inserter(smtDomain),
             typeToSort);
   FnDecl decl(smtDomain, typeToSort(range), string(name) + "_tvfn");
-  return DeclaredFunction(move(domain), move(range), move(decl));
+
+  vector<FnDecl> dims;
+  if (auto sty = range.dyn_cast<mlir::ShapedType>()) {
+    const auto rank = sty.getRank();
+    dims.reserve(rank);
+    const auto dimPrefix = string(name) + "_tvfn_dim_";
+    for (size_t i = 0; i < rank; i++) {
+      auto dim = FnDecl(smtDomain, Index::sort(), dimPrefix + to_string(i));
+      dims.push_back(move(dim));
+    }
+  }
+  return DeclaredFunction(move(domain), move(range), move(decl), move(dims));
 }
 
 ValueTy DeclaredFunction::apply(const std::vector<ValueTy> &operands) const {
   vector<Expr> operandExprs;
   operandExprs.reserve(operands.size());
 
+  auto getZeroGuardedExpr = [](const ValueTy &val) {
+    if (holds_alternative<Tensor>(val)) {
+      const auto tensorVal = get<Tensor>(val);
+      Expr numElements = Index::one();
+      for (const auto &dim : tensorVal.getDims()) {
+        numElements = numElements * dim;
+      }
```

Comment:
I guess there is a function for doing this as well somewhere.

Reply:
I first thought of using `std::accumulate`, but `Expr::operator*` is not a binary operator. I also looked for methods like `getNumElements` or sth like that but couldn't find it...

Reply:
I found one - it is `get1DSize` at smt.h. :) You can use it.

Reply:
oooooh I see

Reply:
Fixed